src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -297,168 +354,117 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-                try
-                {
-                    ContentLengthMethodSetter.Invoke(httpWebRequest, new object[] { contentLength });
-                    headerSet = true;
-                }
-                catch (InvalidOperationException)
-                {
-                    SettingContentLengthPropertyValidOnCurrentPlatform = false;
-                }
-            }
+                _requestMessage.Content = new ByteArrayContent(messageContent);
 
-            if (!headerSet && SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform)
-            {
-                // Attempting to set the header but may just not work
-                try
-                {
-                    // Need this code to set the user agent on silverlight when the property doesn't exist
-                    httpWebRequest.Headers[XmlConstants.HttpContentLength] = XmlConvert.ToString(contentLength);
-                }
-                catch (ArgumentException)
+                _messageStream.Dispose();
+
+                // Apply cached "Content" header values
+                foreach (KeyValuePair<string, string> contentHeader in _contentHeaderValueCache)
                 {
-                    SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform = false;
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
                 }
             }
-#else
-            httpWebRequest.ContentLength = contentLength;
-#endif
+
+            _requestMessage.Method = new HttpMethod(_effectiveHttpMethod);
+
+            return _client.SendAsync(_requestMessage);
         }
 
-        /// <summary>
-        /// Sets the accept charset.
-        /// </summary>
-        /// <param name="httpWebRequest">The HTTP web request.</param>
-        /// <param name="headerValue">The header value.</param>
-        internal static void SetAcceptCharset(HttpWebRequest httpWebRequest, string headerValue)
+        private static IDictionary<string, string> HttpHeadersToStringDictionary(HttpHeaders headers)
         {
-#if !PORTABLELIB
-            httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-#endif
-#if PORTABLELIB
-            if (!IsRunningOnSilverlight)
-            {
-                httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-            }
-#endif
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
         }
 
-        /// <summary>
-        /// Attempts to set the UserAgent property if it exists other wise returns false
-        /// </summary>
-        /// <param name="httpWebRequest">The http web request.</param>
-        /// <param name="headerValue">The value of the user agent.</param>
-        internal static void SetUserAgentHeader(HttpWebRequest httpWebRequest, string headerValue)
+        /// <summary> Returns the underlying HttpRequestMessage </summary>
+        public HttpRequestMessage HttpRequestMessage

```

Comment:
@ElizabethOkerio Doesn't this make this PR a possible breaking change?
We previously had the following:
        ```
        /// <summary> Returns the underlying HttpWebRequest </summary>
        public System.Net.HttpWebRequest HttpWebRequest
        {
            get
            {
                return this.httpRequest;
            }
        }
```
I don't know what the impact would be but it does look like a person who was relying that property would get into trouble
cc. @General

Reply:
I think this will probably need to be documented when the change goes live.